README
#GlgLineGraph (a.k.a cairo version of glinegraph )  July 2007/2016

![GLineGraph Widget](https://github.com/skoona/glinegraph-cairo/raw/master/glg_cairo.png) 

    A Gtk3/GLib2/cario GUI application which demonstrates the use of GTK+ for
 producing xy line graphs.  This widget once created allows you to add one or more
 data series, then add values to those series for ploting.  The X point is assumed based
 on arrival order.  However, the Y value or position is based one the current scale and
 the y value itself.  If the charts x scale maximum is 40, or 40 points, the 41+ value is
 appended to the 40th position after all points are shifted left and out through pos 0.
 
 Packaged as a gtk widget for ease of use.	

 A lgcairo.c example program is included to demonstrate the possible use of the codeset.

There is also gtk-style API documentation in ./src/docs/reference/html/index.html

FEATURES: 
    * GlgLineGraph API Reference Manual in ./src/docs/reference/html/index.html
    * Unlimited data series support.
    * Accurate scaling across a wide range of X & Y scales.
      - Using values ranges above or below 1.
    * Rolling data points, if number of x points exceed x-scale. (left shift)
    * Ability to change chart background color, window backgrounds colors, etc.
    * Popup Tooltip, via mouse-button-1 click to enable/toggle.
      - Tooltip overlays top graph title, when present.
    * Data points are time stamped with current time when added.
	* Some key debug messages to console with button-3 enable, then button-2 click/toggle.
	* Auto Size to current window size; i.e. no-scrolling.
	* point-selected signal tied to tooltip, to display y value under mouse.

REQUIREMENTS:

	Gtk3/Glib2 runtime support. (Created on Fedora Core 6)

	* the following packages may need to be installed to build program and
	  for runtime of binaries on some platforms.

	glinegraph - { this package }
	gtk-devel  - { GTK+ version 3.10 is minimum package level required}
	glib-devel - { version 2.40, packaged with GTK+ } 
						 		

DISTRIBUTION METHOD:

	Source tar	glinegraph-{version}.tar.bz2
	GPL2 copyright

INSTALL INFO:
	Configure Source with 
			'# ./autogen.sh '
			'# make '
			'#sudo make install'

INSTRUCTIONS: 

  glinegraph -- GTK+ Cairo Programing Example Application
    
    1. Compile program
	2. Execute sample program
		- regular cmd: "# lgcairo" 


KNOWN BUGS:
 Gtk+-3.10.0 has a major performance bug when drawing with Cairo.  Researching the Failure now.
 First noticed on OSX, then Raspbian, Fedora, and Ubuntu.
 Related to 'Draw' signal and cairo surface used for painting: first pass 4ms, each pass adds 300ms more
 until the time to draw is greater than the arrival on new data points, causing a draw backlog.



BEGIN DEBUG LOG: to see console log -- $ export G_MESSAGES_DEBUG=all

[jscott@imac glinegraph-cairo]$ src/lgcairo 
** (lgcairo:6489): DEBUG: ===> glg_line_graph_class_init()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_init()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_chart_set_color()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_chart_set_color()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_chart_set_color()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_chart_set_color()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_chart_set_text()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_chart_set_text()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_set_property()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_chart_set_text()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_configure_event(entered)
** (lgcairo:6489): DEBUG: ===> glg_line_graph_configure_event(new width=540, height=260)
** (lgcairo:6489): DEBUG: cfg:factors:raw:pango_layout_get_pixel_size(width=10, height=12)
** (lgcairo:6489): DEBUG: cfg:factors:adj:pango_layout_get_pixel_size(width=10, height=20)
** (lgcairo:6489): DEBUG: cfg:Max.Avail: plot_box.width=470, plot_box.height=165
** (lgcairo:6489): DEBUG: cfg:Chart:Incs:    x_minor=9, x_major=45, y_minor=3, y_major=15, plot_box.x=76, plot_box.y=50, plot_box.width=450, plot_box.height=165
** (lgcairo:6489): DEBUG: cfg:Chart:Nums:    x_num_minor=50, x_num_major=10, y_num_minor=55, y_num_major=11
** (lgcairo:6489): DEBUG: cfg:Chart:Plot:    x=76, y=50, width=450, height=165
** (lgcairo:6489): DEBUG: cfg:Chart:Title:   x=76, y=5, width=450, height=40
** (lgcairo:6489): DEBUG: cfg:Chart:yLabel:  x=5, y=215, width=30, height=140
** (lgcairo:6489): DEBUG: cfg:Chart:xLabel:  x=76, y=230, width=450, height=25
** (lgcairo:6489): DEBUG: cfg:Chart:Tooltip: x=76, y=5, width=450, height=45
** (lgcairo:6489): DEBUG: ===> glg_line_graph_configure_event(exited)
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_add()
** (lgcairo:6489): DEBUG:   ==>DataSeriesAdd: series=0, max_pts=100
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_add()
** (lgcairo:6489): DEBUG:   ==>DataSeriesAdd: series=1, max_pts=100
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_add()
** (lgcairo:6489): DEBUG:   ==>DataSeriesAdd: series=2, max_pts=100
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_add()
** (lgcairo:6489): DEBUG:   ==>DataSeriesAdd: series=3, max_pts=100
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_add()
** (lgcairo:6489): DEBUG:   ==>DataSeriesAdd: series=4, max_pts=100
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_add_value()
** (lgcairo:6489): DEBUG:   ==>DataSeriesAddValue: series=0, value=7.7, index=0, count=1, max_pts=100
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_add_value()
** (lgcairo:6489): DEBUG:   ==>DataSeriesAddValue: series=1, value=19.5, index=0, count=1, max_pts=100
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_add_value()
** (lgcairo:6489): DEBUG:   ==>DataSeriesAddValue: series=2, value=82.7, index=0, count=1, max_pts=100
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_add_value()
** (lgcairo:6489): DEBUG:   ==>DataSeriesAddValue: series=3, value=58.7, index=0, count=1, max_pts=100
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_add_value()
** (lgcairo:6489): DEBUG:   ==>DataSeriesAddValue: series=4, value=100.0, index=0, count=1, max_pts=100
** (lgcairo:6489): DEBUG: ===> glg_line_graph_redraw()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_add_value()
** (lgcairo:6489): DEBUG:   ==>DataSeriesAddValue: series=0, value=7.7, index=0, count=2, max_pts=100
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_add_value()
** (lgcairo:6489): DEBUG:   ==>DataSeriesAddValue: series=1, value=19.5, index=0, count=2, max_pts=100
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_add_value()
** (lgcairo:6489): DEBUG:   ==>DataSeriesAddValue: series=2, value=82.7, index=0, count=2, max_pts=100
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_add_value()
** (lgcairo:6489): DEBUG:   ==>DataSeriesAddValue: series=3, value=19.3, index=0, count=2, max_pts=100
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_add_value()
** (lgcairo:6489): DEBUG:   ==>DataSeriesAddValue: series=4, value=100.0, index=0, count=2, max_pts=100
** (lgcairo:6489): DEBUG: ===> glg_line_graph_redraw()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_expose(entered)
** (lgcairo:6489): DEBUG: glg_line_graph_expose(width=540, height=260)
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw(entered)
** (lgcairo:6489): DEBUG: Chart.Surface: pg.Width=540, pg.Height=260, Plot Area x=76 y=50 width=450, height=165
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw_text_horizontal()
** (lgcairo:6489): DEBUG: Horiz.TextBox:Page cx=540, cy=260
** (lgcairo:6489): DEBUG: Horiz.TextBox:Orig: x=76, y=5, cx=450, cy=40
** (lgcairo:6489): DEBUG: Horiz.TextBox:Calc x_pos=187, y_pos=8,  cx=228, cy=36
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw_text_horizontal()
** (lgcairo:6489): DEBUG: Horiz.TextBox:Page cx=540, cy=260
** (lgcairo:6489): DEBUG: Horiz.TextBox:Orig: x=76, y=230, cx=450, cy=25
** (lgcairo:6489): DEBUG: Horiz.TextBox:Calc x_pos=155, y_pos=237,  cx=291, cy=16
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw_text_vertical()
** (lgcairo:6489): DEBUG: Vert:TextBox: y_pos=239,  x=5, y=215, cx=218, cy=30
** (lgcairo:6489): DEBUG: Vert.TextBox: y_pos=239,  x=5, y=215, cx=218, cy=30
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw_grid_lines()
** (lgcairo:6489): DEBUG: Draw.Y-GridLines: count_major=10, count_minor=54, y_minor_inc=3, y_major_inc=15
** (lgcairo:6489): DEBUG: Draw.X-GridLines: count_major=9, count_minor=49, x_minor_inc=9, x_major_inc=45
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw_x_grid_labels()
** (lgcairo:6489): DEBUG: Scale:Labels:X small font sizes cx=20, cy=11
** (lgcairo:6489): DEBUG: Scale:Labels:X plot_box.cx=450, layout.cx=470, layout.cy=11
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw_y_grid_labels()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_draw_all(entered)
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_draw()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_draw()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_draw()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_draw()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_draw()
** (lgcairo:6489): DEBUG: glg_line_graph_data_series_draw_all(exited): #series=5
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw_tooltip()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw(exited)
** (lgcairo:6489): DEBUG: glg_line_graph_expose(exited)
** (lgcairo:6489): DEBUG: ===> glg_line_graph_configure_event(entered)
** (lgcairo:6489): DEBUG: ===> glg_line_graph_configure_event(new width=540, height=260)
** (lgcairo:6489): DEBUG: cfg:factors:raw:pango_layout_get_pixel_size(width=10, height=12)
** (lgcairo:6489): DEBUG: cfg:factors:adj:pango_layout_get_pixel_size(width=10, height=20)
** (lgcairo:6489): DEBUG: cfg:Max.Avail: plot_box.width=470, plot_box.height=165
** (lgcairo:6489): DEBUG: cfg:Chart:Incs:    x_minor=9, x_major=45, y_minor=3, y_major=15, plot_box.x=76, plot_box.y=50, plot_box.width=450, plot_box.height=165
** (lgcairo:6489): DEBUG: cfg:Chart:Nums:    x_num_minor=50, x_num_major=10, y_num_minor=55, y_num_major=11
** (lgcairo:6489): DEBUG: cfg:Chart:Plot:    x=76, y=50, width=450, height=165
** (lgcairo:6489): DEBUG: cfg:Chart:Title:   x=76, y=5, width=450, height=40
** (lgcairo:6489): DEBUG: cfg:Chart:yLabel:  x=5, y=215, width=30, height=140
** (lgcairo:6489): DEBUG: cfg:Chart:xLabel:  x=76, y=230, width=450, height=25
** (lgcairo:6489): DEBUG: cfg:Chart:Tooltip: x=76, y=5, width=450, height=45
** (lgcairo:6489): DEBUG: ===> glg_line_graph_configure_event(exited)
** (lgcairo:6489): DEBUG: ===> glg_line_graph_expose(entered)
** (lgcairo:6489): DEBUG: glg_line_graph_expose(width=540, height=260)
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw(entered)
** (lgcairo:6489): DEBUG: Chart.Surface: pg.Width=540, pg.Height=260, Plot Area x=76 y=50 width=450, height=165
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw_text_horizontal()
** (lgcairo:6489): DEBUG: Horiz.TextBox:Page cx=540, cy=260
** (lgcairo:6489): DEBUG: Horiz.TextBox:Orig: x=76, y=5, cx=450, cy=40
** (lgcairo:6489): DEBUG: Horiz.TextBox:Calc x_pos=187, y_pos=8,  cx=228, cy=36
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw_text_horizontal()
** (lgcairo:6489): DEBUG: Horiz.TextBox:Page cx=540, cy=260
** (lgcairo:6489): DEBUG: Horiz.TextBox:Orig: x=76, y=230, cx=450, cy=25
** (lgcairo:6489): DEBUG: Horiz.TextBox:Calc x_pos=155, y_pos=237,  cx=291, cy=16
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw_text_vertical()
** (lgcairo:6489): DEBUG: Vert:TextBox: y_pos=239,  x=5, y=215, cx=218, cy=30
** (lgcairo:6489): DEBUG: Vert.TextBox: y_pos=239,  x=5, y=215, cx=218, cy=30
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw_grid_lines()
** (lgcairo:6489): DEBUG: Draw.Y-GridLines: count_major=10, count_minor=54, y_minor_inc=3, y_major_inc=15
** (lgcairo:6489): DEBUG: Draw.X-GridLines: count_major=9, count_minor=49, x_minor_inc=9, x_major_inc=45
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw_x_grid_labels()
** (lgcairo:6489): DEBUG: Scale:Labels:X small font sizes cx=20, cy=11
** (lgcairo:6489): DEBUG: Scale:Labels:X plot_box.cx=450, layout.cx=470, layout.cy=11
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw_y_grid_labels()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_draw_all(entered)
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_draw()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_draw()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_draw()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_draw()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_data_series_draw()
** (lgcairo:6489): DEBUG: glg_line_graph_data_series_draw_all(exited): #series=5
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw_tooltip()
** (lgcairo:6489): DEBUG: ===> glg_line_graph_draw(exited)
** (lgcairo:6489): DEBUG: glg_line_graph_expose(exited)

END DEBUG LOG:
 	


NEWS
* Thur May 26, 2016   Release 0.4.2
- Migrated to GTK+-3.0
- contains: (glg) widget style routines for easier integration
- fixed draw performance problem caused by '360 * M_PI'	    

* Wed Jul 04, 2007   Release 0.3.1-0
- Major rewrite of widget components
- contains: (glg) widget style routines for easier integration
		    enhanced with Cairo Graphics Engine support

* Tue Jul 03, 2007   Release 0.3.0-0
- Major rewrite of widget components
- contains: (glg) widget style routines for easier integration
		    considering an enhancement to support Cairo Engine
            
* Wed May 24, 2006   Release 0.2.0-0
- First public release of glinegraph package
- contains: (g_lgraph) regular routines for direct integration
            (sklinegraph) widget style routines for easier integration



autogen.sh
#!/bin/sh
# Run this to generate all the initial makefiles, etc.

srcdir=`dirname $0`
test -z "$srcdir" && srcdir=.

(test -f $srcdir/configure.ac) || {
        echo "**Error**: Directory "\`$srcdir\'" does not look like the top-level project directory"
        exit 1
}

PKG_NAME=`autoconf --trace "AC_INIT:$1" "$srcdir/configure.ac"`

if [ "$#" = 0 -a "x$NOCONFIGURE" = "x" ]; then
        echo "**Warning**: I am going to run \`configure' with no arguments." >&2
        echo "If you wish to pass any to it, please specify them on the" >&2
        echo \`$0\'" command line." >&2
        echo "" >&2
fi

gtkdocize || exit 1

# if the AC_CONFIG_MACRO_DIR() macro is used, create that directory
# This is a automake bug fixed in automake 1.13.2
# See http://debbugs.gnu.org/cgi/bugreport.cgi?bug=13514
m4dir=`autoconf --trace 'AC_CONFIG_MACRO_DIR:$1'`
if [ -n "$m4dir" ]; then
        mkdir -p $m4dir
fi

set -x

#intltoolize --force --copy --automake || exit 1
autoreconf --verbose --force --install -Wno-portability || exit 1

if [ "$NOCONFIGURE" = "" ]; then
        $srcdir/configure "$@" || exit 1

        if [ "$1" = "--help" ]; then exit 0 else
                echo "Now type \`make\' to compile $PKG_NAME" || exit 1
        fi
else
        echo "Skipping configure process."
fi

set +x


src/glg_cairo.h
/* $Id: glg_cairo.h,v 1.35 2007/07/25 16:41:07 jscott Exp $
 * ----------------------------------------------
 *
 * A GTK+ widget that implements a XY line graph
 *
 * (c) 2007, 2016 James Scott Jr 
 *
 * Authors:
 *   James Scott Jr <skoona@gmail.com>
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */


#ifndef __GLG_LINE_GRAPH_H__
#define __GLG_LINE_GRAPH_H__

G_BEGIN_DECLS

/**
 * Constants:
 * @GLG_USER_MODEL_X: Minimum graph width before auto-scaling.
 * @GLG_USER_MODEL_Y: Minimum graph height before auto-scaling.
 */
#define GLG_USER_MODEL_X 570     /* Minimum width */
#define GLG_USER_MODEL_Y 270     /* Minimum height */

/**
 * @GLG_MAX_STRING: Maximum gchar string size for any api.
 */
#define GLG_MAX_STRING  256      /* Size of a text string */

typedef struct _GlgLineGraph  GlgLineGraph;
typedef struct _GlgLineGraphClass GlgLineGraphClass;
typedef struct _GlgLineGraphPrivate GlgLineGraphPrivate;


/**
 * GlgLineGraphClass:
 * @point-selected: signal to return the y value under or near the mouse.
 * @graph: pointer to a #GlgLineGraph widget
 * @x_value: x scale value
 * @y_value: y scale value
 * @point_y_pos: y value pixel position on chart
 * @mouse_y_pos: actual mouse y position on chart
 *
 * Main widget Class structure
 */
struct _GlgLineGraphClass
{
    GtkWidgetClass parent_class;

	void	(* point_selected)	(GlgLineGraph *graph, double x_value, double y_value, double point_y_pos, double mouse_y_pos);
    /* Padding for future expansion */
    void (*_glg_reserved1) (void);
    void (*_glg_reserved2) (void);
    void (*_glg_reserved3) (void);
    void (*_glg_reserved4) (void);
};

/**
 * GlgLineGraph:
 *
 * Main widget structure
 */
struct _GlgLineGraph
{
	GtkWidget parent;

	/* < private > */
	GlgLineGraphPrivate *priv;
};


/**
 * GLGElementID:
 * @GLG_TITLE_X:     Enables display of the bottom chart title
 * @GLG_NO_TITLE_X:  Disables display of the bottom chart title
 * @GLG_TITLE_Y:    Enables display of the left/vertical chart title
 * @GLG_NO_TITLE_Y: Disables display of the left/vertical chart title
 * @GLG_TITLE_T:    Enables display of the top chart title
 * @GLG_NO_TITLE_T: Disables display of the top chart title
 * @GLG_GRID_LABELS_X:
 * @GLG_NO_GRID_LABELS_X:
 * @GLG_GRID_LABELS_Y:
 * @GLG_NO_GRID_LABELS_Y:
 * @GLG_TOOLTIP:
 * @GLG_NO_TOOLTIP:
 * @GLG_GRID_LINES:
 * @GLG_NO_GRID_LINES:
 * @GLG_GRID_MINOR_X:
 * @GLG_NO_GRID_MINOR_X:
 * @GLG_GRID_MAJOR_X:
 * @GLG_NO_GRID_MAJOR_X:
 * @GLG_GRID_MINOR_Y:
 * @GLG_NO_GRID_MINOR_Y:
 * @GLG_GRID_MAJOR_Y:
 * @GLG_NO_GRID_MAJOR_Y:
 * @GLG_SCALE:   chart color key -- used to change chart scale/labels color
 * @GLG_TITLE:   chart color key -- used to change top title color
 * @GLG_WINDOW:  chart color key -- used to change window color
 * @GLG_CHART:   chart color key -- used to change chart color
 * 
 * Communication params for interface APIs
 */
typedef enum _GLG_Graph_Elements {
  /* enable chart flags and title keys */
  GLG_ELEMENT_NONE = 1 << 0,
  GLG_TITLE_X = 1 << 1,
  GLG_NO_TITLE_X = 0 << 1,
  GLG_TITLE_Y = 1 << 2,
  GLG_NO_TITLE_Y = 0 << 2,
  GLG_TITLE_T = 1 << 3,
  GLG_NO_TITLE_T = 0 << 3,

  /* enable chart attributes flag */  
  GLG_GRID_LABELS_X = 1 << 4,
  GLG_NO_GRID_LABELS_X = 0 << 4,
  GLG_GRID_LABELS_Y = 1 << 5,
  GLG_NO_GRID_LABELS_Y = 0 << 5,

  /* enable tooltip flag */
  GLG_TOOLTIP = 1 << 6,
  GLG_NO_TOOLTIP = 0 << 6,  
  
  /* enabled chart attributes */
  GLG_GRID_LINES   = 1 << 7,
  GLG_NO_GRID_LINES   = 0 << 7,
  GLG_GRID_MINOR_X = 1 << 8,
  GLG_NO_GRID_MINOR_X = 0 << 8,
  GLG_GRID_MAJOR_X = 1 << 9,
  GLG_NO_GRID_MAJOR_X = 0 << 9,
  GLG_GRID_MINOR_Y = 1 << 10,
  GLG_NO_GRID_MINOR_Y = 0 << 10,
  GLG_GRID_MAJOR_Y = 1 << 11,
  GLG_NO_GRID_MAJOR_Y = 0 << 11,
  
  /* chart color key -- used to change window color only */
  GLG_SCALE   = 1 << 12,  
  GLG_TITLE   = 1 << 13,
  GLG_WINDOW  = 1 << 14,  
  GLG_CHART   = 1 << 15, 

  /* Reserved */   
  GLG_RESERVED_OFF = 0 << 16,
  GLG_RESERVED_ON  = 1 << 16      
} GLGElementID;

#define GLG_TYPE_LINE_GRAPH			(glg_line_graph_get_type ())
#define GLG_LINE_GRAPH(obj)			(G_TYPE_CHECK_INSTANCE_CAST ((obj), GLG_TYPE_LINE_GRAPH, GlgLineGraph))
#define GLG_IS_LINE_GRAPH(obj)		(G_TYPE_CHECK_INSTANCE_TYPE ((obj), GLG_TYPE_LINE_GRAPH))


/* 
 * Public Interfaces 
*/
extern GlgLineGraph * glg_line_graph_new (const gchar *first_property_name, ...);
extern GType    glg_line_graph_get_type (void) G_GNUC_CONST;
extern void 		glg_line_graph_redraw (GlgLineGraph *graph);
extern GLGElementID glg_line_graph_chart_get_elements ( GlgLineGraph *graph);
extern void 		glg_line_graph_chart_set_elements ( GlgLineGraph *graph, GLGElementID element);
extern gboolean 	glg_line_graph_chart_set_text  (GlgLineGraph *graph, GLGElementID element, const gchar *pch_text);
extern gboolean 	glg_line_graph_chart_set_color (GlgLineGraph *graph, GLGElementID element, const gchar *pch_color);  
extern void 		glg_line_graph_chart_set_ranges (GlgLineGraph *graph,
                                 gint x_tick_minor, gint x_tick_major,
                                 gint x_scale_min,  gint x_scale_max,
                                 gint y_tick_minor, gint y_tick_major, 
                                 gint y_scale_min,  gint y_scale_max);
extern void glg_line_graph_chart_set_x_ranges (GlgLineGraph *graph,
                                 gint x_tick_minor, gint x_tick_major,
                                 gint x_scale_min,  gint x_scale_max);
extern void glg_line_graph_chart_set_y_ranges (GlgLineGraph *graph,
                                 gint y_tick_minor, gint y_tick_major,
                                 gint y_scale_min,  gint y_scale_max);


extern gint 		glg_line_graph_data_series_add (GlgLineGraph *graph, const gchar *pch_legend_text, const gchar *pch_color_text);
extern gboolean 	glg_line_graph_data_series_add_value (GlgLineGraph *graph, gint i_series_number, gdouble y_value);

G_END_DECLS

#endif


src/lgcairo.c
/* lgcairo.c
 * ---------------------------------------------------------------
 * Test glg Line Graph Widget in a GtkWindow
 *
 * (c) 2007, 2016 James Scott Jr
 * Date: 5/2016
 *
 * Authors:
 *   James Scott Jr <skoona@gmail.com>
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include <gtk/gtk.h>

#include "glg_cairo.h"


static gint gb_stop = FALSE;

void fn_show_point_selected (GtkWidget *graph, double x_value,     double y_value, 
												  double point_y_pos, double mouse_y_pos, gpointer data)
{
	g_print ("fn_show_point_selected: x_value=%1.0lf, y_value=%1.1lf, y_pos=%1.1lf, mouse_y_pos=%1.1lf, ranking=%1.1lf\n",
			  x_value, y_value, point_y_pos, mouse_y_pos, 
			  (point_y_pos < mouse_y_pos) ? ((point_y_pos / mouse_y_pos) * 100) : (( mouse_y_pos / point_y_pos) * 100) ); 
	return;
}
/*
 * Test Update
 *
 */
static gint fn_add_series_data_values (GlgLineGraph *graph)
{
	if ( gb_stop ) return FALSE;

    g_return_val_if_fail (graph != NULL, TRUE);

    if(GTK_IS_WIDGET(graph) == FALSE) {
        return TRUE;
    }


    glg_line_graph_data_series_add_value (graph, 0, g_random_double_range (5.0, 20.0) );
    glg_line_graph_data_series_add_value (graph, 1, g_random_double_range (20.0, 30.0) );
    glg_line_graph_data_series_add_value (graph, 2, 82.7 );
    glg_line_graph_data_series_add_value (graph, 3, g_random_double_range (30.0, 82.0) );
    glg_line_graph_data_series_add_value (graph, 4, g_random_double_range (98.0, 100.0) );

    glg_line_graph_redraw ( graph );

    return (TRUE);
}


/*
 * to enable debug messages
 * $ export G_MESSAGES_DEBUG=all
 */
int main (int argc, char **argv)
{
	GtkWidget *window = NULL;
	GlgLineGraph *glg;
	guint  ui_add_values = 0; 

	gtk_init (&argc, &argv);

	window = gtk_window_new (GTK_WINDOW_TOPLEVEL);
	gtk_window_set_title (GTK_WINDOW(window), "GlgLineGraph Widget Example"); 
	gtk_window_set_default_size (GTK_WINDOW(window), GLG_USER_MODEL_X, GLG_USER_MODEL_Y);
	g_signal_connect (window, "destroy",
					  G_CALLBACK (gtk_main_quit), NULL);

        /*
         * IMPORTANT
         * set chart x and y increments and ranges
         * it is important to make tick-minor-x as small as possible or 1
         * and to also limit the scale-major-x to a value less than 100
         * otherwise the chart gets too busy.
         *
         * Also can only be set once, recreate graph widget to change this value
         */

	/*
	 * Create Graph -- Alternate Init Method
	 * and set all its options in one call */
	glg = glg_line_graph_new ("chart-set-elements", GLG_TOOLTIP | GLG_TITLE_T | GLG_TITLE_X | GLG_TITLE_Y | GLG_GRID_MAJOR_X |  GLG_GRID_MAJOR_Y | GLG_GRID_MINOR_X |  GLG_GRID_MINOR_Y | GLG_GRID_LABELS_X | GLG_GRID_LABELS_Y,
			"range-tick-minor-x", 1,
			"range-tick-major-x", 10,
			"range-scale-minor-x", 0,
			"range-scale-major-x", 40,
			"range-tick-minor-y", 2,
			"range-tick-major-y", 10,
			"range-scale-minor-y", 0,
			"range-scale-major-y", 110,
			"series-line-width", 2,
			"graph-title-foreground",  "blue",
			"graph-scale-foreground",  "black",
			"graph-chart-background",  "light blue",
			"graph-window-background", "white",
			"text-title-main", "<big><b>A GTK Line Graph Widget</b></big>\n<span foreground=\"orange\"><b><i>using Cairo Graphics</i></b></span>",
			"text-title-yaxis", "Pango Text Markup\n<small>is supported in all <span foreground=\"red\">X &amp; Y</span> titles.</small>",
			"text-title-xaxis", "<i>Click mouse button 1 to <span foreground=\"red\">toggle</span> popup legend.</i>",
			NULL);

	gtk_container_add (GTK_CONTAINER (window), GTK_WIDGET(glg));
          

	/*
	 * setup to catch tooltip drill downs 
	 */
    g_signal_connect (glg, "point-selected", G_CALLBACK(fn_show_point_selected),NULL);

 /* 
  * Add empty data series containers, then add data later via timer routine 
  * this routine returns an int which is the series-number for plotting data
  * it should be retained and used with the glg_line_graph_data_series_add_value() api
  * the values are returned in sequence from 0 to N-series */

    glg_line_graph_data_series_add (glg, "Volts", "red");
    glg_line_graph_data_series_add (glg, "Battery", "blue");
    glg_line_graph_data_series_add (glg, "Load", "green");
    glg_line_graph_data_series_add (glg, "Charge", "orange");
    glg_line_graph_data_series_add (glg, "Line", "yellow");

	ui_add_values = g_timeout_add (1000, (GSourceFunc) fn_add_series_data_values, glg);

	gtk_widget_show_all (window);

	gtk_main ();
	
	gb_stop = TRUE;
	
	g_source_remove (ui_add_values );
	 
	return 0;
}
